nano/node/active_transactions.cpp

```diff
@@ -297,16 +336,121 @@ void nano::active_transactions::frontiers_confirmation (nano::unique_lock<std::m
 		lock_a.unlock ();
 		auto transaction = node.store.tx_begin_read ();
-		prioritize_frontiers_for_confirmation (transaction, node.network_params.network.is_dev_network () ? std::chrono::milliseconds (50) : time_spent_prioritizing_ledger_accounts, time_spent_prioritizing_wallet_accounts);
-		confirm_prioritized_frontiers (transaction);
+		prioritize_frontiers_for_confirmation (transaction, node.network_params.network.is_dev_network () ? std::chrono::milliseconds (50) : time_to_spend_prioritizing_ledger_accounts, time_to_spend_prioritizing_wallet_accounts);
+		auto frontiers_confirmation_info = get_frontiers_confirmation_info ();
+		if (frontiers_confirmation_info.can_start_elections ())
+		{
+			uint64_t elections_count (0);
+			confirm_prioritized_frontiers (transaction, frontiers_confirmation_info.max_elections, elections_count);
+			confirm_expired_frontiers_pessimistically (transaction, frontiers_confirmation_info.max_elections, elections_count);
+			set_next_frontier_check (frontiers_confirmation_info.aggressive_mode);
+		}
 		lock_a.lock ();
 	}
 }
 
+void nano::active_transactions::confirm_expired_frontiers_pessimistically (nano::transaction const & transaction_a, uint64_t max_elections_a, uint64_t & elections_count_a)
+{
+	auto i{ node.store.latest_begin (transaction_a, next_frontier_account) };
+	auto n{ node.store.latest_end () };
+	nano::timer<std::chrono::milliseconds> timer (nano::timer_state::started);
+	nano::confirmation_height_info confirmation_height_info;
+	uint64_t elections_count{ 0 };
+
+	for (auto i = expired_optimistic_elections.get<tag_expired_time> ().begin (); i != expired_optimistic_elections.get<tag_expired_time> ().end ();)
+	{
+		if (stopped || elections_count_a >= max_elections_a)
+		{
+			break;
+		}
+
+		auto const & account{ i->account };
+		nano::account_info account_info;
+		bool should_delete{ true };
+		if (!node.store.account_get (transaction_a, account, account_info) && !node.store.confirmation_height_get (transaction_a, account, confirmation_height_info))
+		{
+			if (account_info.block_count > confirmation_height_info.height)
+			{
+				should_delete = false;
+				std::shared_ptr<nano::block> previous_block;
+				std::shared_ptr<nano::block> block;
+				if (confirmation_height_info.height == 0)
+				{
+					block = node.store.block_get (transaction_a, account_info.open_block);
+				}
+				else
+				{
+					previous_block = node.store.block_get (transaction_a, confirmation_height_info.frontier);
+					block = node.store.block_get (transaction_a, previous_block->sideband ().successor);
+				}
+
+				if (block && !node.confirmation_height_processor.is_processing_block (block->hash ()))
+				{
+					auto source{ block->source () };
+					if (source.is_zero ())
+					{
+						source = block->link ();
+					}
+
+					auto should_start_election{ false };
+
+					// Check if the source is confirmed (if open/receive) before trying to confirm this block
```

Comment:
So I originally didn't use it because we already know that the previous dependent is already confirmed (comes from cemented frontier), so it seems a waste to check it again. However it does better encapsulate the logic. Rather than try to change it to fit the needs here I'm just gonna use it. Changed name to `ledger::dependents_confirmed` and also replaced contents with `block_confirmed` (could possibly use `block_confirmed_or_confirming` too but this at least matches behaviour which is currently there). Thanks.